doppelganger/doppelganger.go

```diff
@@ -0,0 +1,265 @@
+		slotTickerProvider: opts.SlotTickerProvider,
+		logger:             opts.Logger.Named(logging.NameDoppelganger),
+		doppelgangerState:  make(map[phase0.ValidatorIndex]*doppelgangerState),
+	}
+}
+
+// ValidatorStatus returns the current status of the validator.
+func (ds *doppelgangerHandler) ValidatorStatus(validatorIndex phase0.ValidatorIndex) DoppelgangerStatus {
+	ds.mu.RLock()
+	defer ds.mu.RUnlock()
+
+	state, exists := ds.doppelgangerState[validatorIndex]
+	if !exists {
+		ds.logger.Warn("Validator not found in Doppelganger state", fields.ValidatorIndex(validatorIndex))
+		return UnknownToDoppelganger
+	}
+
+	if state.requiresFurtherChecks() {
+		return SigningDisabled
+	}
+	return SigningEnabled
+}
+
+func (ds *doppelgangerHandler) MarkAsSafe(validatorIndex phase0.ValidatorIndex) {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+
+	state, exists := ds.doppelgangerState[validatorIndex]
+	if !exists {
+		ds.logger.Warn("Validator not found in Doppelganger state", fields.ValidatorIndex(validatorIndex))
+		return
+	}
+
+	ds.logger.Debug("mark validator as doppelganger safe", fields.ValidatorIndex(validatorIndex))
+	state.remainingEpochs = 0
+}
+
+func (ds *doppelgangerHandler) updateDoppelgangerState(validatorIndices []phase0.ValidatorIndex) {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+
+	// Create a set for current validators
+	currentValidatorSet := make(map[phase0.ValidatorIndex]struct{}, len(validatorIndices))
+	for _, idx := range validatorIndices {
+		currentValidatorSet[idx] = struct{}{}
+	}
+
+	// Add new validators with DefaultRemainingDetectionEpochs
+	for _, idx := range validatorIndices {
+		if _, exists := ds.doppelgangerState[idx]; !exists {
+			ds.doppelgangerState[idx] = &doppelgangerState{
+				remainingEpochs: DefaultRemainingDetectionEpochs,
+			}
+			ds.logger.Debug("Added validator to Doppelganger state", fields.ValidatorIndex(idx))
+		}
+	}
+
+	// Remove validators that are no longer in the current set
+	for idx := range ds.doppelgangerState {
+		if _, exists := currentValidatorSet[idx]; !exists {
+			ds.logger.Debug("Removing validator from Doppelganger state", fields.ValidatorIndex(idx))
+			delete(ds.doppelgangerState, idx)
+		}
+	}
+}
+
+func (ds *doppelgangerHandler) RemoveValidatorState(validatorIndex phase0.ValidatorIndex) {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+
+	if _, exists := ds.doppelgangerState[validatorIndex]; !exists {
+		ds.logger.Warn("Validator not found in Doppelganger state", fields.ValidatorIndex(validatorIndex))
+		return
+	}
+
+	delete(ds.doppelgangerState, validatorIndex)
+	ds.logger.Debug("Removed validator from Doppelganger state", fields.ValidatorIndex(validatorIndex))
+}
+
+func (ds *doppelgangerHandler) StartMonitoring(ctx context.Context) {
+	ds.logger.Info("Doppelganger monitoring started")
+
+	firstRun := true
+	ticker := ds.slotTickerProvider()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.Next():
+			currentSlot := ticker.Slot()
```

Comment:
slot ticker can skip slots forward or backward due to system clock, system sleep or high load

we should make sure to keep validators disabled under these scenarios

let's focus on slot skipping forward or triggering the same slot twice, as these are the most likely fault conditions

the simplest solution i can think of is to keep a `previousEpoch` above the loop and if it didn't go up exactly by 1 from the last iteration: we should reset all doppelganger states back to default (2 remaining epochs)

(checking `previousEpoch` should be done after the last epoch trigger)

wdyt?

Reply:
Maybe we can track `previousSlot` instead of `previousEpoch` (or both). If `currentSlot` jumps forward more than expected (e.g., more than 1 slot at a time), we could temporarily pause doppelganger checks until normal slot progression has resumed.